From c1122fd25c291a279f348d4063547b18671b4a63 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Mon, 13 Feb 2017 17:36:37 +0200
Subject: [PATCH 0836/1345] fix: net: phy: m88e1112: fix 10Mbits autoneg off
 wrong state issue

commit  5abc9e834604b46fa75dca27036822affac3b58f from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Issue:
- PHY 88E1112 in mode speed 10Mbits&autoneg off doesn't recognize
  link partner speed change and stuck with link UP state while
  link partner link state changed to DOWN.

Fix:
- Reset phy if energy state changed in 10Mbits autoneg off state.

Change-Id: I2e6cd30d1aa703ae0c98638e1e4beeefb3b4541a
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37167
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/phy/marvell.c |   20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/drivers/net/phy/marvell.c b/drivers/net/phy/marvell.c
index 0ff2d21..562d146 100644
--- a/drivers/net/phy/marvell.c
+++ b/drivers/net/phy/marvell.c
@@ -86,6 +86,8 @@
 #define MII_M1111_COPPER		0
 #define MII_M1111_FIBER			1
 
+#define MII_M1112_PHY_ENERGY_STATE	0x10
+
 #define MII_88E1121_PHY_MSCR_PAGE	2
 #define MII_88E1121_PHY_MSCR_REG	21
 #define MII_88E1121_PHY_MSCR_RX_DELAY	BIT(5)
@@ -1891,6 +1893,22 @@ static int m88e1510_probe(struct phy_device *phydev)
 	return m88e1510_hwmon_probe(phydev);
 }
 
+static int m88e1112_read_status(struct phy_device *phydev)
+{
+	int val, bmcr;
+
+	if ((phydev->autoneg != AUTONEG_ENABLE) & (phydev->speed == SPEED_10)) {
+		val = phy_read(phydev, MII_M1011_IEVENT);
+		/* PHY reset required to detect link partner speed change and proper status read */
+		if (val & MII_M1112_PHY_ENERGY_STATE) {
+			bmcr = phy_read(phydev, MII_BMCR);
+			phy_write(phydev, MII_BMCR, bmcr | BMCR_RESET);
+		}
+	}
+
+	return genphy_read_status(phydev);
+}
+
 static struct phy_driver marvell_drivers[] = {
 	{
 		.phy_id = MARVELL_PHY_ID_88E1101,
@@ -1919,7 +1937,7 @@ static int m88e1510_probe(struct phy_device *phydev)
 		.probe = marvell_probe,
 		.config_init = &m88e1111_config_init,
 		.config_aneg = &marvell_config_aneg,
-		.read_status = &genphy_read_status,
+		.read_status = &m88e1112_read_status,
 		.ack_interrupt = &marvell_ack_interrupt,
 		.config_intr = &marvell_config_intr,
 		.resume = &genphy_resume,
-- 
1.7.9.5

